sdk/core/azure-core/src/cryptography/md5.cpp

```diff
@@ -16,100 +16,66 @@
-            sizeof(hashLength),
-            &dataLength,
-            0);
-        if (!BCRYPT_SUCCESS(status))
-        {
-          throw std::runtime_error("BCryptGetProperty failed");
-        }
-        HashLength = hashLength;
-      }
-
-      ~AlgorithmProviderInstance() { BCryptCloseAlgorithmProvider(Handle, 0); }
-    };
-
-    struct Md5HashContext
-    {
-      std::string buffer;
-      BCRYPT_HASH_HANDLE hashHandle = nullptr;
-      std::size_t hashLength = 0;
-    };
-  } // namespace _detail
+struct AlgorithmProviderInstance
+{
+  BCRYPT_ALG_HANDLE Handle;
+  std::size_t ContextSize;
+  std::size_t HashLength;
 
-  Md5Hash::Md5Hash()
+  AlgorithmProviderInstance()
   {
-    static _detail::AlgorithmProviderInstance AlgorithmProvider{};
-
-    _detail::Md5HashContext* md5Context = new _detail::Md5HashContext;
-    m_md5Context = md5Context;
-    md5Context->buffer.resize(AlgorithmProvider.ContextSize);
-    md5Context->hashLength = AlgorithmProvider.HashLength;
-
-    NTSTATUS status = BCryptCreateHash(
-        AlgorithmProvider.Handle,
-        &md5Context->hashHandle,
-        reinterpret_cast<PUCHAR>(&md5Context->buffer[0]),
-        static_cast<ULONG>(md5Context->buffer.size()),
-        nullptr,
-        0,
+    NTSTATUS status = BCryptOpenAlgorithmProvider(&Handle, BCRYPT_MD5_ALGORITHM, nullptr, 0);
+    if (!BCRYPT_SUCCESS(status))
+    {
+      throw std::runtime_error("BCryptOpenAlgorithmProvider failed");
+    }
+    DWORD objectLength = 0;
+    DWORD dataLength = 0;
+    status = BCryptGetProperty(
+        Handle,
+        BCRYPT_OBJECT_LENGTH,
+        reinterpret_cast<PBYTE>(&objectLength),
+        sizeof(objectLength),
+        &dataLength,
         0);
     if (!BCRYPT_SUCCESS(status))
     {
-      throw std::runtime_error("BCryptCreateHash failed");
+      throw std::runtime_error("BCryptGetProperty failed");
+    }
+    ContextSize = objectLength;
+    DWORD hashLength = 0;
+    status = BCryptGetProperty(
+        Handle,
+        BCRYPT_HASH_LENGTH,
+        reinterpret_cast<PBYTE>(&hashLength),
+        sizeof(hashLength),
+        &dataLength,
+        0);
+    if (!BCRYPT_SUCCESS(status))
+    {
+      throw std::runtime_error("BCryptGetProperty failed");
     }
+    HashLength = hashLength;
   }
 
-  Md5Hash::~Md5Hash()
-  {
-    _detail::Md5HashContext* md5Context = static_cast<_detail::Md5HashContext*>(m_md5Context);
-    BCryptDestroyHash(md5Context->hashHandle);
-    delete md5Context;
-  }
+  ~AlgorithmProviderInstance() { BCryptCloseAlgorithmProvider(Handle, 0); }
+};
 
-  void Md5Hash::OnAppend(const uint8_t* data, std::size_t length)
-  {
-    _detail::Md5HashContext* md5Context = static_cast<_detail::Md5HashContext*>(m_md5Context);
+class Md5BCrypt : public Azure::Core::Cryptography::Hash {
+private:
+  std::string m_buffer;
+  BCRYPT_HASH_HANDLE m_hashHandle = nullptr;
+  std::size_t m_hashLength = 0;
+  AlgorithmProviderInstance m_algorithmProviderInstance;
```

Comment:
I've got another question now. Are we going to create a new `AlgorithmProviderInstance` instance for each MD5 calculation?

Reply:
Yes, there are two size_t fields in it and a handle.  There's not a super difference from using only one.
The benefit, in the other hand, of letting each md5 instance to have it's own is to avoid someone updating the global instance at some point and changing all md5 instances right?  For instance, if someone re-asign it for SHA or something else, all md5 instances would react to it.  It would be a weird edge case, but we can prevent it.

Reply:
This is not good. The old impl is all MD5 instances share the same algorithm handle, because in the API reference it says

```
the BCryptOpenAlgorithmProvider function is a relatively time intensive function. Because of this, we recommend that you cache any algorithm provider handles that you will use more than once, rather than opening and closing the algorithm providers over and over.
```

[link](https://docs.microsoft.com/en-us/windows/win32/api/bcrypt/nf-bcrypt-bcryptopenalgorithmprovider)

Reply:
> For instance, if someone re-asign it for SHA or something else, all md5 instances would react to it.

Different algorithm will have different static handles, so that they won't interfere with each other.

Reply:
oh crap. Well, I will update it to use only one provider then. Thank you @Jinming-Hu